Implement Client.getWeight() and add RunEnergyPlugin

seandewar opened the pull request:
This plugin shows the following tooltip when the user mouses over the run energy orb:
![image](https://user-images.githubusercontent.com/6256228/43973331-1b66f4f6-9ccf-11e8-8f0a-a98f9447e072.png)

And has the option to replace the run energy percentage next to the orb with the remaining run time:
![](https://i.imgur.com/vdQtMTB.png)

This plugin gives an estimated time until run energy has been depleted and until run energy recharges to 100%. It also shows the user's weight (as shown in the equipment screen).

The plugin considers the Graceful set effect (with the Agility and Max Cape substitution) and stamina boosts. Weights under 0 kg have no effect on the estimated depletion time as expected.

Closes #4666.

MagicfTail commented on runelite-client/src/main/java/net/runelite/client/plugins/runenergy/RunEnergyOverlay.java:
```diff
@@ -0,0 +1,225 @@
+		GRACEFUL_CAPE_13593, GRACEFUL_CAPE_13594, GRACEFUL_CAPE_13605, GRACEFUL_CAPE_13606,
+		GRACEFUL_CAPE_13617, GRACEFUL_CAPE_13618, GRACEFUL_CAPE_13629, GRACEFUL_CAPE_13630,
+		GRACEFUL_CAPE_13669, GRACEFUL_CAPE_13670, GRACEFUL_CAPE_21064, GRACEFUL_CAPE_21066,
+		AGILITY_CAPE, AGILITY_CAPET, AGILITY_CAPE_10648, AGILITY_CAPE_13340, AGILITY_CAPET_13341
```
Also max cape

seandewar replied:
Good catch! This only includes the non-cosmetic versions, right? (e.g. doesn't include the max fire cape, for example)

MagicfTail replied:
Ye. Also, why are there 42 different agility capes? Are you sure some of them aren't mounted versions or something?

seandewar replied:
Is there any way to check?

MagicfTail replied:
Don't think so no, but it looks like `AGILITY_CAPE` and `AGILITY_CAPET` are the correct items

seandewar replied:
Okay, cheers! I'm also not sure about the huge amount of Graceful item variations but I'm unsure of how to check those. :(

MagicfTail replied:
Ignore that comment lol

Kamielvf commented on runelite-client/src/main/java/net/runelite/client/plugins/runenergy/RunEnergyOverlay.java:
```diff
@@ -0,0 +1,226 @@
+		if (isWearingFullGraceful())
+		{
+			// 30% recover rate increase from Graceful set effect
+			recoverRate *= 1.3;
```
Graceful doesn't work like this. It only applies the effect if you are standing still

seandewar replied:
Cheers. I'll have a look in a sec. :)

seandewar replied:
If I understand correctly, run energy only ever regenerates when the player's not running anyway (and not busy doing something such as crossing agility obstacles).

In this case, maybe it would be useful to hide the recovery time while running?

seandewar commented:
I've changed things so that the recovery time displays N/A when running to a destination (moving faster than 1 tile per tick with a waypoint set).

Also, I've added the config ability to replace the run orb text with the seconds of run remaining.

I might reword the tooltip contents to make it less verbose and maybe add a color shift to the tooltip text when less than 15s of run energy remains.

Let me know what you think. :)

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/runenergy/RunEnergyPlugin.java:
```diff
@@ -0,0 +1,250 @@
+public class RunEnergyPlugin extends Plugin
+{
+	// TODO It would be nice if we have the IDs for just the equipped variants of the Graceful set items.
+	private static final ImmutableSet ALL_GRACEFUL_HOODS = ImmutableSet.of(
```
`ImmutableSet<Integer>`

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/runenergy/RunEnergyPlugin.java:
```diff
@@ -0,0 +1,250 @@
+			runOrbIcon.getSpriteId() == SpriteID.MINIMAP_ORB_RUN_ICON_SLOWED_DEPLETION);
+	}
+
+	public String getEstimatedRunTimeRemaining(boolean inSeconds)
```
this should be package private

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/runenergy/RunEnergyPlugin.java:
```diff
@@ -0,0 +1,250 @@
+			ALL_GRACEFUL_CAPES.contains(items[EquipmentInventorySlot.CAPE.getSlotIdx()].getId()));
+	}
+
+	public String getEstimatedRecoverTimeRemaining()
```
package private

seandewar commented:
Changes added. Also changed the recovery time left text to be less verbose. :)

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/runenergy/RunEnergyOverlay.java:
```diff
@@ -0,0 +1,87 @@
+			tooltipManager.add(new Tooltip(tooltip));
+		}
+
+		return new Dimension((int) bounds.getWidth(), (int) bounds.getHeight());
```
You can just return bounds.getSize() or simply null as this is not layoutable overlay.

seandewar replied:
Thanks. I've fixed this. 👍

ShaunDreclin commented:
👀👌🏻
